### PR TITLE
Pinot fixes

### PR DESCRIPTION
**Pull request by elonazoulay**
Fixes to pinot-connector:
* Fix count(*) passthrough queries (passthrough meaning the query is in the table name)
* Allow functions and other expressions in passthrough queries
* Limit pinot segment queries - pinot should not be used for large selects
* Fix column mapping for passthrough queries: the mapping was inverted

### CODE REVIEW

**Comment by elonazoulay** [comment]
Note: will add a warning for pinot segment queries where the limit was not applied or was greater than the configured maximum and apply the limit regardless instead of failing the query.

File: presto-pinot/src/main/java/io/prestosql/pinot/PinotBrokerPageSource.java [hunk 1]
```diff
@@ -126,6 +128,9 @@ public Page getNextPage()
             blocks[i] = columnBuilders[i].build();
             columnBuilders[i] = columnBuilders[i].newBlockBuilderLike(null);
         }
+        if (decoders.isEmpty()) {
+            return new Page(rowCount);
```
Comment (martint) [inline presto-pinot/src/main/java/io/prestosql/pinot/PinotBrokerPageSource.java:132]: Can we add a test for this bug? Also, add an explanation to the commit message describing what was the issue.

File: presto-pinot/src/main/java/io/prestosql/pinot/query/FilterToPqlConverter.java [hunk 2]
```diff
@@ -215,11 +214,11 @@ private static void checkOperator(FilterQuery filterQuery, FilterOperator filter
 
     private String getPinotColumnName(String columnName)
     {
-        String pinotColumnName = ((PinotColumnHandle) columnHandles.get(columnName)).getColumnName();
-        if (pinotColumnName == null) {
-            throw new ColumnNotFoundException(schemaTableName, columnName);
+        PinotColumnHandle columnHandle = ((PinotColumnHandle) columnHandles.get(columnName));
```
Comment (martint) [inline presto-pinot/src/main/java/io/prestosql/pinot/query/FilterToPqlConverter.java:217]: Please add a test

File: presto-pinot/src/main/java/io/prestosql/pinot/client/PinotClient.java [hunk 3]
```diff
@@ -479,12 +479,12 @@ private BrokerResponseNative submitBrokerQueryJson(ConnectorSession session, Pin
     public Iterator<BrokerResultRow> createResultIterator(ConnectorSession session, PinotQuery query, List<PinotColumnHandle> columnHandles)
     {
         BrokerResponseNative response = submitBrokerQueryJson(session, query);
-        Map<String, Integer> columnIndices = getColumnIndices(columnHandles);
         ResultTable resultTable = response.getResultTable();
         String[] columnNames = resultTable.getDataSchema().getColumnNames();
+        Map<String, Integer> columnIndices = getColumnIndices(columnNames);
```
Comment (martint) [inline presto-pinot/src/main/java/io/prestosql/pinot/client/PinotClient.java:484]: Please add a test

File: presto-pinot/src/test/java/io/prestosql/pinot/util/TestingPinotCluster.java [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+        checkState(response.getStatus().equals(format("%s successfully added", tableName)), "Unexpected response: '%s'", response.getStatus());
+        verifySchema(tableName);
+    }
+
+    private URI getControllerUri(String path)
+    {
+        return URI.create(format("http://%s/%s", getControllerConnectString(), path));
+    }
+
+    public void verifySchema(String tableName)
```
Comment (martint) [inline presto-pinot/src/test/java/io/prestosql/pinot/util/TestingPinotCluster.java:158]: private?

File: presto-pinot/src/test/java/io/prestosql/pinot/util/TestingPinotCluster.java [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+            throws Exception
+    {
+        Request request = Request.builder().prepareGet().setUri(getControllerUri("schemas"))
```
Comment (martint) [inline presto-pinot/src/test/java/io/prestosql/pinot/util/TestingPinotCluster.java:161]: `prepareGet()` is a static method on Request.Builder, so do this instead: `Request.Builder.prepareGet()`

This applies to all other usages of prepareXXX

File: presto-pinot/src/test/java/io/prestosql/pinot/util/TestingPinotCluster.java [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+        zookeeper.start();
+        controller.start();
+        broker.start();
+        server.start();
+    }
+
+    @Override
+    public void close()
+    {
+        server.stop();
```
Comment (martint) [inline presto-pinot/src/test/java/io/prestosql/pinot/util/TestingPinotCluster.java:111]: The http client should also be closed here.

File: presto-pinot/src/main/java/io/prestosql/pinot/PinotPageSourceProvider.java [hunk 7]
```diff
@@ -41,18 +40,18 @@
     private final PinotConfig pinotConfig;
     private final PinotQueryClient pinotQueryClient;
     private final PinotClient clusterInfoFetcher;
-    private final ObjectMapper objectMapper;
+    private final int limitForSegmentQueries;
```
Comment (martint) [inline presto-pinot/src/main/java/io/prestosql/pinot/PinotPageSourceProvider.java:42]: How is this related to this commit?

File: presto-pinot/src/test/java/io/prestosql/pinot/TestMinimalFunctionality.java [hunk 8]
```diff
@@ -16,31 +16,38 @@
...
 import java.util.List;
+import java.util.Map;
+import java.util.Optional;
 
 import static io.airlift.testing.Closeables.closeAllRuntimeException;
 import static java.util.Objects.requireNonNull;
 
 @Test(singleThreaded = true)
 public class TestMinimalFunctionality
+        extends AbstractTestQueryFramework
```
Comment (martint) [inline presto-pinot/src/test/java/io/prestosql/pinot/TestMinimalFunctionality.java:42]: How is this related to this commit?

File: presto-pinot/src/main/java/io/prestosql/pinot/PinotPlugin.java [hunk 9]
```diff
@@ -14,15 +14,37 @@
...
+import io.prestosql.pinot.client.PinotHostMapper;
 import io.prestosql.spi.Plugin;
 import io.prestosql.spi.connector.ConnectorFactory;
 
+import static java.util.Objects.requireNonNull;
+
 public class PinotPlugin
         implements Plugin
 {
+    public static final Module DEFAULT_EXTENSION = binder -> {
```
Comment (martint) [inline presto-pinot/src/main/java/io/prestosql/pinot/PinotPlugin.java:29]: `{ }` are unnecessary for single-expression lambda.

File: presto-pinot/src/main/java/io/prestosql/pinot/PinotPlugin.java [hunk 10]
```diff
@@ -14,15 +14,37 @@
...
+        binder.bind(PinotHostMapper.class).to(IdentityPinotHostMapper.class).in(Scopes.SINGLETON);
```
Comment (martint) [inline presto-pinot/src/main/java/io/prestosql/pinot/PinotPlugin.java:30]: It'd probably be cleaner to user Guice's OptionalBinder for this.

File: presto-pinot/src/main/java/io/prestosql/pinot/query/PinotQueryBuilder.java [hunk 11]
```diff
@@ -71,15 +72,13 @@ public static String generatePql(PinotTableHandle tableHandle, List<PinotColumnH
...
-        OptionalLong limit = tableHandle.getLimit();
-        if (limit.isPresent()) {
-            pqlBuilder.append(" LIMIT ")
-                    .append(limit.getAsLong());
-        }
-        else {
-            pqlBuilder.append(" LIMIT ")
-                    .append(Integer.MAX_VALUE);
+        OptionalLong appliedLimit = tableHandle.getLimit();
+        long limit = limitForSegmentQueries;
```
Comment (martint) [inline presto-pinot/src/main/java/io/prestosql/pinot/query/PinotQueryBuilder.java:67]: If I understand correctly, when the config property is set, all queries will get a LIMIT injected. Is that right? 

If so, this is problematic, as it might cause queries to silently return fewer results than the user expected. The correct behavior would be to fail if the query produces more rows than the configured limit. To do so, we can inject `limit + 1`, and if the count of rows returns equals `limit + 1`, fail the query.

File: presto-pinot/src/main/java/io/prestosql/pinot/PinotSegmentPageSource.java [hunk 12]
```diff
@@ -366,8 +371,8 @@ Block getArrayBlock(int rowIndex, int columnIndex)
             case FLOAT_ARRAY:
                 float[] floatArray = currentDataTable.getDataTable().getFloatArray(rowIndex, columnIndex);
                 blockBuilder = elementType.createBlockBuilder(null, floatArray.length);
-                for (double element : floatArray) {
-                    elementType.writeDouble(blockBuilder, element);
+                for (float element : floatArray) {
+                    elementType.writeLong(blockBuilder, floatToIntBits(element));
```
Comment (martint) [inline presto-pinot/src/main/java/io/prestosql/pinot/PinotSegmentPageSource.java:373]: This should be `writeInt()`

File: presto-pinot/src/main/java/io/prestosql/pinot/decoders/RealDecoder.java [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+import static io.prestosql.spi.type.RealType.REAL;
+import static java.lang.Float.floatToIntBits;
+
+public class RealDecoder
+        implements Decoder
+{
+    @Override
+    public void decode(Supplier<Object> getter, BlockBuilder output)
+    {
+        REAL.writeLong(output, floatToIntBits(((Number) getter.get()).floatValue()));
```
Comment (martint) [inline presto-pinot/src/main/java/io/prestosql/pinot/decoders/RealDecoder.java:29]: `writeInt()`

File: presto-pinot/src/main/java/io/prestosql/pinot/PinotSegmentPageSource.java [hunk 14]
```diff
@@ -193,7 +191,7 @@ private void fetchPinotData()
 
             this.columnTypes = columnHandles
                     .stream()
-                    .map(columnHandle -> getTypeForBlock(columnHandle))
+                    .map(columnHandle -> columnHandle.getDataType())
```
Comment (martint) [inline presto-pinot/src/main/java/io/prestosql/pinot/PinotSegmentPageSource.java:194]: Can you add a test for integer types?

File: presto-pinot/src/main/java/io/prestosql/pinot/query/DynamicTableBuilder.java [hunk 15]
```diff
@@ -42,7 +42,7 @@
     private static final String COLUMN_KEY = "column";
```
Comment (martint) [inline presto-pinot/src/main/java/io/prestosql/pinot/query/DynamicTableBuilder.java:42]: Add a short explanation to the commit message describing what was the issue.

File: pom.xml [hunk 16]
```diff
@@ -1217,6 +1217,13 @@
                 <artifactId>rubix-presto-shaded</artifactId>
                 <version>${rubix.version}</version>
             </dependency>
+
+            <dependency>
+                <groupId>io.prestosql</groupId>
+                <artifactId>presto-kafka</artifactId>
+                <version>${project.version}</version>
+                <type>test-jar</type>
+            </dependency>
```
[the remark this answers is not in the document]
Comment (elonazoulay) [reply pom.xml:1226]: I wanted to reuse TestingKafka, I can use the testing kafka from testcontainers directly.

File: presto-pinot/src/test/java/io/prestosql/pinot/TestMinimalFunctionality.java [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+        pinot.start();
+        createTopic();
+        produceRecords();
+        addSchema();
+        addRealtime();
+
+        Map<String, String> pinotProperties = ImmutableMap.<String, String>builder()
+                .put("pinot.controller-urls", pinot.getControllerConnectString())
+                .build();
+        return PinotQueryRunner.createPinotQueryRunner(ImmutableMap.of(), pinotProperties, Optional.of(binder -> {
```
[the remark this answers is not in the document]
Comment (elonazoulay) [reply presto-pinot/src/test/java/io/prestosql/pinot/TestMinimalFunctionality.java:65]: AbstractTestQueryFramework.createQueryRunner returns QueryRunner, compiles without throwing an error.

File: presto-pinot/src/main/java/io/prestosql/pinot/query/PinotQueryBuilder.java (hunk 11 above)
Comment (elonazoulay) [reply presto-pinot/src/main/java/io/prestosql/pinot/query/PinotQueryBuilder.java:67]: Sounds good, have to do this per split as there is no way that I'm aware of to keep track of the global rowcount from the page source.

File: pom.xml (hunk 16 above)
[the remark this answers is not in the document]
Comment (elonazoulay) [reply pom.xml]: Moved to presto-testing.

**Comment by elonazoulay** [comment]
Depends on https://github.com/prestosql/tempto/pull/55 to resolve the class not found error in product tests.

**Comment by elonazoulay** [comment]
Hey @martin, this is ready for review, let me know how it looks.

**Comment by elonazoulay** [comment]
Separated into #4722, #4723, #4724, #4725 and #4726
